pkg/sparrow/run.go

```diff
@@ -83,35 +94,32 @@ func New(cfg *config.Config) *Sparrow {
 
 // Run starts the sparrow
 func (s *Sparrow) Run(ctx context.Context) error {
-	ctx, cancel := logger.NewContextWithLogger(ctx, "sparrow")
-	log := logger.FromContext(ctx)
+	ctx, cancel := logger.NewContextWithLogger(ctx)
 	defer cancel()
 
-	go s.loader.Run(ctx)
-
-	if s.targets != nil {
-		go s.targets.Reconcile(ctx)
-	}
-
-	// Start the api
 	go func() {
-		err := s.api(ctx)
-		if err != nil {
-			log.Error("Error running api server", "error", err)
+		s.cErr <- s.loader.Run(ctx)
+	}()
+	go func() {
+		if s.targets != nil {
+			s.cErr <- s.targets.Reconcile(ctx)
 		}
 	}()
+	go func() {
+		s.cErr <- s.api(ctx)
+	}()
+
+	go s.handleErrors(ctx)
 
 	for {
 		select {
-		case <-ctx.Done():
-			return s.shutdown(ctx)
 		case result := <-s.cResult:
 			go s.db.Save(result)
 		case configChecks := <-s.cCfgChecks:
-			// Config got updated
-			// Set checks
 			s.cfg.Checks = configChecks
 			s.ReconcileChecks(ctx)
+		case <-s.cDone:
+			return fmt.Errorf("sparrow was shut down")
 		}
 	}
```

Comment:
Why not move the handleErrors function into the select?

Reply:
Yeah, that was overengineering. Thanks for catching that one - we only care about the context errors and in case a non-recoverable error happens, so I just removed the function and added the two cases in the select. This was an unnecessary indirection.

Having two extra cases in the control loop of the `Run` function is clearer.